arch/arc/core/arc_smp.c

```diff
@@ -103,18 +72,60 @@ void z_arch_slave_start(int cpu_num)
 {
 	void (*fn)(int, void*);
 
+#ifdef CONFIG_SMP
 	z_icache_setup();
 	z_irq_setup();
 
 	z_irq_priority_set(IRQ_ICI, ARCV2_ICI_IRQ_PRIORITY, 0);
 	irq_enable(IRQ_ICI);
-
+#endif
 	/* call the function set by z_arch_start_cpu */
 	fn = arc_cpu_init[cpu_num].fn;
 
 	fn(cpu_num, arc_cpu_init[cpu_num].arg);
 }
 
+#ifdef CONFIG_SMP
+
+static void sched_ipi_handler(void *unused)
+{
+	ARG_UNUSED(unused);
+
+	z_arc_connect_ici_clear();
+	z_sched_ipi();
+}
+
+/**
+ * @brief Check whether need to do thread switch in isr context
+ *
+ * @details u64_t is used to let compiler use (r0, r1) as return register.
+ *  use register r0 and register r1 as return value, r0 has
+ *  new thread, r1 has old thread. If r0 == 0, it means no thread switch.
+ */
+u64_t z_arch_smp_switch_in_isr(void)
+{
+	u64_t ret = 0;
+	u32_t new_thread;
+	u32_t old_thread;
+
+	old_thread = (u32_t)_current;
+
+	new_thread = (u32_t)z_get_next_ready_thread();
+
+	if (new_thread != old_thread) {
+#ifdef CONFIG_TIMESLICING
+		z_reset_time_slice();
+#endif
```

Comment:
This is another spot where you're doing something the scheduler should have already done for you.  Why is the time slice code needed here?

Reply:
I think it may be required. each time when there is a thread switch , z_reset_time_slice should be called.   For SMP's case  z_reset_time_slice() will not be called in update_cache like single processor.

so consider a case of ipi interrupt.   One thread T2 is is made ready by T1 in core A through  z_add_thread_to_ready_q and will run in core B through ipi interrupt.  As there are no more higher priority ready  thread to run , then T1 in core A will continue to run, z_reset_time_slice will not be calledd because no swap in core A.   For core B, if  z_reset_time_slice is not called, i think there will be problem.